### ISSUE TITLE
Update action regex can match / delete unintended lines on CentOS 7

### ISSUE THREAD

**Issue by detjensrobert** [karnauskas/chef-yum-plugin-versionlock#27]
## :bug: Problem

**TLDR: If a package has dashes in its name, the current regex used by `:update` can match multiple package names and delete unintended locks on Centos 7 and below.**

We are using this cookbook to lock some Docker packages on Centos 7, namely `docker-ce` and `docker-ce-cli`. We have two versionlock resources `:update`ing the lock for each of those packages.

The update regex is `/^[0-9]+:$PKGNAME-.+-.+\./`, which when filled in for `docker-ce` becomes `^[0-9]+:docker-ce-.+-.+\.` The `-.+` directly following the package name matches the `-cli` of the `docker-ce-cli` lockfile line instead of the version. This would cause all matching lines--both `docker-ce` and `docker-ce-cli`--to be replaced with the `docker-ce` version string, and as of #25 any duplicate lines are removed. The line for `docker-ce-cli` is then re-added by its lock resource. 

The end result is valid, but as the `-cli` line is removed and readded each converge, this breaks idempotency.

On Centos 8, the epoch and `:` following the package name prevents this from happening.

We are in the process of updating our internal cookbooks to be idempotent and as such are just noticing this now. I believe this may have also been present with the old `Chef::FileEdit`-based resource as well, since that used the same regex.

## :bulb: Possible Solution

Updating the regex to look for a number at the start of the version string may fix this, but I am not sure if that would hold for all packages. Some might still have a number after a dash in their name and still trigger this issue, but this would solve this issue for most packages.

Proposed C6/7 regex:
```regex
/^\d+:$PKGNAME-\d.*-.+\./
```

I can make a pull request to make this change, but I wanted to see if there is a better way of handling this first.

**Comment by aleksandrs-ledovskis** [comment]
There doesn't seem to be a proper Regexp way to match package by its _NEVRA_ string (reverse is possible).

Even Yum itself does a `sprintf`-type matching https://github.com/rpm-software-management/yum/blob/1222f377b8dcecb77456ea378a28e1dc23ba4207/yum/packageSack.py#L289-L311

Yes, there is [a regexp](https://github.com/rpm-software-management/yum/blob/1222f377b8dcecb77456ea378a28e1dc23ba4207/test/skipbroken-tests.py#L7) in the Yum repo test section, but it doesn't help here much.

---

I think that a stable anchor could be some kind special [locklist comment](https://github.com/rpm-software-management/yum-utils/blob/master/plugins/versionlock/versionlock.py#L54-L55) above the package's _NEVRA_ reference.

E.g.,

```
#
# File is managed by Chef. 
# Edits will be reverted upon client run!
#

# chef-yum-plugin-versionlock docker-ce
3:docker-ce-19.03.13-3.el7
# chef-yum-plugin-versionlock docker-ce-cli
3:docker-ce-cli-19.03.13-3.el7
```

**Comment by karnauskas** [comment]
Some people, when confronted with a problem, think “I know, I'll use regular expressions.”   Now they have two problems. :)

**Comment by karnauskas** [comment]
Isn't it easier and more transparent to make cookbook to use cli tool, which is doing this already, so no need to worry about regex formats?

**Comment by detjensrobert** [comment]
Unfortunately, no.
The command-line tool requires the desired version of the package to already be installed, as it only takes the name as input and pulls the version info from the installed package. Editing the file directly allows for locking a package prior to installing it, as all of the info can be specified.
